_posts/2019-01-10-future-workloads.md

```diff
@@ -0,0 +1,500 @@
+2. Make small programs communicate to build larger ones. (blocking, asynchronous)
+3. Use all the CPU's in our machines. (cpu bound)
+4. Enable human-computer interaction (GUIs, terminal). (asynchronous)
+
+#### CPU bound
+
+Let's fill up the thread pool and consequentially our cores with actual work.
+
+```
+Future.sequence(List.fill(1000)(Future(cpuBoundMethod(1000)))).await()
+```
+
+On my machine, this produces the following output.
+
+```
+Thread[scala-execution-context-global-11,5,main]
+Thread[scala-execution-context-global-13,5,main]
+Thread[scala-execution-context-global-10,5,main]
+Thread[scala-execution-context-global-12,5,main]
+~1000ms PAUSE
+Thread[scala-execution-context-global-10,5,main]
+Thread[scala-execution-context-global-11,5,main]
+Thread[scala-execution-context-global-12,5,main]
+Thread[scala-execution-context-global-13,5,main]
+~1000ms PAUSE
+Thread[scala-execution-context-global-11,5,main]
+Thread[scala-execution-context-global-10,5,main]
+Thread[scala-execution-context-global-12,5,main]
+Thread[scala-execution-context-global-13,5,main]
+... continues until 1000 executions
+```
+
+And `htop` will consistently show my cpu usage as such:
+
+```
+1  [||||||||||||||||||||||||||||100.0%]
+2  [||||||||||||||||||||||||||||100.0%]
+3  [||||||||||||||||||||||||||||100.0%]
+4  [||||||||||||||||||||||||||||100.0%]
+```
+
+We can see that our cores are filling up with work, but this does not mean our code *performs* better. Although this can be the case, what we may be trying to achieve using asynchronous code is *scalability*.
+
+Creating a future and scheduling it has a cost, and there is a threshold for when it is worth to do this. After this threshold, the concept of amortization kicks in and we start to obtain scaling benefits.
+
+Simplistically, if creating a future and scheduling it takes **10ms**, and our workload takes **10ms** to finish, then it will most likely perform worse. But if our workload takes **1000ms**, the cost of creating the future is amortized in the total cost, and it will scale.
+
+Let's look at an example. The following runs the `cpuBoundMethod` function 4 times **sequentially** in the main thread.
+
+```
+val time: Int = ???
+(1 to 4).foreach(cpuBoundMethod(time))
+```
+
+And using Future instead:
+
+```
+Future.sequence(List.fill(4)(Future(cpuBoundMethod(time)))).await()
+```
+
+My approximate simplistic results (in ms):
+
+|           ms| Sequential | Future |
+|-------------|------------|--------|
+| time = 20   | 80         | 173    |
+| time = 40   | 160        | 198    |
+| time = 60   | 240        | 191    |
+| time = 500  | 2000       | 594    |
+| time = 1000 | 4000       | 1103   |
+
+At 20ms, Future is too costly, At 40ms, Future begins to hit the threshold. At 500ms and 1000ms it scales better.
+
+The above is actually a case of [Amdahl's law](https://en.wikipedia.org/wiki/Amdahl%27s_law). The serial part would be the overhead of creating Future in the main thread.
+
+## Blocking
```

Comment:
headings levels look inconsistent  
article starts with 4th level then jumps to 2nd

Reply:
nice catch. It should be using 4th and 5th only throughout the post. fixed these